packages/grid/_modules_/grid/hooks/utils/useGridApiEventHandler.ts

```diff
@@ -12,30 +15,26 @@ export enum GridSignature {
   DataGridPro = 'DataGridPro',
 }
 
-export function useGridApiEventHandler<E extends GridEventsStr>(
-  apiRef: GridApiRef,
-  eventName: E,
-  handler?: GridEventListener<E>,
-  options?: EventListenerOptions,
-) {
-  const subscription = React.useRef<(() => void) | null>(null);
-  const handlerRef = React.useRef<GridEventListener<E> | undefined>();
-  handlerRef.current = handler;
-
-  if (!subscription.current && handlerRef.current) {
-    const enhancedHandler: GridEventListener<E> = (params, event, details) => {
-      if (!event.defaultMuiPrevented) {
-        handlerRef.current?.(params, event, details);
-      }
-    };
+// We use class to make it easier to detect in heap snapshots by name
+class ObjectToBeRetainedByReact {}
+
+// Based on https://github.com/Bnaya/use-dispose-uncommitted/blob/main/src/finalization-registry-based-impl.ts
+// Check https://github.com/facebook/react/issues/15317 to get more information
+export function createUseGridApiEventHandler(registry: CleanupTracking) {
+  let cleanupTokensCounter = 0;
 
-    subscription.current = apiRef.current.subscribeEvent(eventName, enhancedHandler, options);
-  } else if (!handlerRef.current && subscription.current) {
-    subscription.current();
-    subscription.current = null;
-  }
+  return function useGridApiEventHandler<E extends GridEventsStr>(
+    apiRef: GridApiRef,
+    eventName: E,
+    handler?: GridEventListener<E>,
+    options?: EventListenerOptions,
+  ) {
+    const [objectRetainedByReact] = React.useState(new ObjectToBeRetainedByReact());
```

Comment:
```suggestion
    const [objectRetainedByReact] = React.useState(() => new ObjectToBeRetainedByReact());
```

May be totally negligible though.